### ISSUE TITLE
BUG: Yeo-Johnson Power Transformer gives Numpy warning

### ISSUE THREAD

**Issue by lorentzenchr** [scipy/scipy#18389]
### Describe your issue.

`scipy.stats.yeojohnson` can overflow for reasonable input.

This is the same issue as https://github.com/scikit-learn/scikit-learn/issues/23319.

### Reproducing Code Example

```python
import numpy as np
from scipy.stats import yeojohnson


x = np.array([2003.0, 1950.0, 1997.0, 2000.0, 2009.0, 2009.0, 1980.0, 1999.0, 2007.0, 1991.0])
yeojohnson(x)
```

### Error message

```shell
python3.9/site-packages/numpy/core/_methods.py:239: RuntimeWarning: overflow encountered in multiply
  x = um.multiply(x, x, out=x)
python3.9/site-packages/numpy/core/_methods.py:250: RuntimeWarning: overflow encountered in reduce
  ret = umr_sum(x, axis, dtype, out, keepdims=keepdims, where=where)
Out[4]: 
(array([1.63003810e+154, 4.59874518e+153, 1.41487438e+154, 1.51873102e+154,
        1.87712777e+154, 1.87712777e+154, 9.45185553e+153, 1.48330911e+154,
        1.79094464e+154, 1.22758933e+154]),
 47.211105636582914)
```

### SciPy/NumPy/Python version and system information

```shell
numpy 1.24.1
scipy 1.10.1
```

**Comment by lorentzenchr** [comment]
@lsorber ping

**Comment by WarrenWeckesser** [comment]
Those values *look* reasonable, but in fact, that data set runs into a fundamental limitation of 64 bit floating point when one attempts to find the optimal (i.e. maximum likelihood) Yeo-Johnson parameter.

Using `mpmath`, one can compute the actual optimal parameter $\hat{\lambda}$ to be approximately 99.26.  (I'm adding this to [`mpsci`](https://github.com/WarrenWeckesser/mpsci), but I haven't pushed it to github yet.) The problem is that for each value in that data set, $(x_i + 1)^\hat{\lambda}$ overflows when computed with 64 bit floats. E.g.

```
In [15]: x
Out[15]: 
array([2003., 1950., 1997., 2000., 2009., 2009., 1980., 1999., 2007.,
       1991.])

In [16]: x**99.26
<ipython-input-16-42b558e52fe2>:1: RuntimeWarning: overflow encountered in power
  x**99.26
Out[16]: array([inf, inf, inf, inf, inf, inf, inf, inf, inf, inf])
```
In other words, the result of the optimal Yeo-Johnson transform for this data set is not representable with 64 bit floats.

-----

In practice, a user can avoid this problem by using an additional preprocessing step to scale the data. For example, the data can be normalized to have mean 0 and standard deviation 1 before applying `yeojohnson`. In scikit-learn, one would use `sklearn.preprocessing.StandardScaler`.  In SciPy, `scipy.stats.zscore` can be used:

```
In [23]: from scipy.stats import zscore, yeojohnson

In [24]: z = zscore(x)

In [25]: yeojohnson(z)
Out[25]: 
(array([ 0.68110757, -1.01124976,  0.16177483,  0.39752566,  1.40110131,
         1.40110131, -0.54744521,  0.31375023,  1.13770228, -0.18007565]),
 2.387535792418072)
```

Of course, whether or not such a step is valid for a user depends on why they are transforming the data in the first place and what they are going to do with the transformed data.

**Comment by mdhaber** [comment]
@WarrenWeckesser what do you recommend for this issue? Document this limitation?

**Comment by rkern** [comment]
The proposed fix for `sklearn`'s implementation is here: https://github.com/scikit-learn/scikit-learn/pull/26188

**Comment by lsorber** [comment]
Scikit-learn and scipy have very similar implementations of Yeo-Johnson that are both prone to two sources of overflow:
1. The first source of overflow is caused by certain invocations of `np.power` with unbalanced bases and exponents.
2. The second source of overflow is more difficult and occurs when λ grows for a marginal gain in log likelihood. In this case, applying the Yeo-Johnson transform on the input will cause overflow.

Both types of overflow can be mitigated. The fix that I proposed for scikit-learn (https://github.com/scikit-learn/scikit-learn/pull/26188) resolves them as follows:
1. For the first source of overflow: instances of `np.power` are replaced by instances of `np.exp` so that bases and exponents are combined before being exponentiated.
2. For the second source of overflow: I proposed to regularize the exponents of the transformed input so that they do not blow up for a marginal gain in log likelihood. The regularization term has a small weight relative to the log likelihood objective, and so generally does not affect the result.

The scikit-learn maintainers proposed to fix the issue in scipy and then have scikit-learn depend on its improved implementation.

**Comment by lorentzenchr** [comment]
> The scikit-learn maintainers proposed to fix the issue in scipy and then have scikit-learn depend on its improved implementation.

Yes we did, in the hope to serve the greater community and only have one yeojohnson implementation, i.e. the one in scipy.

I hope the fix/ pr of @lsorber is welcome here.

**Comment by lsorber** [comment]
@WarrenWeckesser 

Item 1: Even better, great!

Item 2: First, a minimal working example that illustrates the problem:
```python
import numpy as np
from scipy.stats import yeojohnson, yeojohnson_llf
x = np.array([2003.0, 1950.0, 1997.0, 2000.0, 2009.0])
xt, lmb = yeojohnson(x)
log_likelihood = yeojohnson_llf(lmb, x)  # RuntimeWarning: overflow encountered in power: out[pos] = (np.power(x[pos] + 1, lmbda) - 1) / lmbda

# xt = array([2.01405528e+154, 5.67790578e+153, 1.74805534e+154, 1.87644716e+154, 2.31954974e+154])
# lmb = 47.23900784138352
# log_likelihood = -13.9893342
```
Notice that the lambda that maximises the log likelihood is rather large, and that the transformed data's average exponent has grown from about 3 to 154. This exponent is already beyond what single precision can represent, and is getting close to what double precision can represent.

By adding a regularisation term that sums the squared exponents of the transformed data `xt` with a weight of 1e-4, we obtain the following result:
```python
# xt = array([349.48615808, 342.69449234, 348.71975305, 349.10303333, 350.25194217])
# lmb = 0.7292862554669863
# log_likelihood = -15.2971991
```
The transformed data are now in a more reasonable range, at the cost of a 10% decrease in log likelihood. The regularisation weight could be zero by default so as not to change the current behaviour. This regularisation term is not standard AFAIK. I proposed it for scikit-learn to address the issues described above.

Another way to deal with the overflow is to change the optimiser from `optimize.brent` to `optimize.fminbound` (which also uses Brent's method) and pre-compute bounds on lambda that would prevent overflow from occurring in the optimisation. This is something I would consider regardless of the decision to add the regularisation term above.

---

Related to this issue: I came across a relatively recent paper [1] that proves that the Yeo-Johnson negative log likelihood objective is convex, and proposes to replace the Brent optimiser with a more robust exponential search. In Figure 2, they also illustrate overflow issues with Brent optimisation of the Yeo-Johnson log likelihood that are similar to the ones described in this issue.

[1] https://arxiv.org/abs/2210.01639

**Comment by mdhaber** [comment]
@lsorber could you submit the PR with at least #1 and the change to `fminbnd`? Perhaps include the regulation idea in a separate commit so it can be considered separately.

**Comment by mdhaber** [comment]
@lsorber @lorentzenchr were you interested in submitting the fix?

**Comment by lsorber** [comment]
Yes, I'd like to help implement the fix according to your proposal. I haven't had much time the past few weeks. Will try to fit it in in the coming weeks!

**Comment by lsorber** [comment]
@WarrenWeckesser @mdhaber @lorentzenchr I implemented fixes (1) and (2) in #18852. These should resolve the two sources of overflow.
